mvc/models/model-binding.rst

```diff
@@ -1,8 +1,89 @@
+ 
+#. ``Form values``: These are form values that go in the HTTP request using the POST method. (including jQuery POST requests).
+#. ``Route values``: The set of route values provided by `routing <https://docs.asp.net/projects/mvc/en/latest/controllers/routing.html>`_. 
+#. ``Query strings``: The query string part of the URI.
+
+.. note:: Form values, route data, and query strings are all stored as name-value pairs.
+
+Since model binding asked for a key named ``id`` and there is nothing named ``id`` in the form values, it moved on to the route values looking for that key. In our example, it's a match. Binding happens, and the value is converted to the integer 2. The same request using Edit(string id) would convert to the string "2". 
+
+So far the example uses primitive types. If the action method's parameter were a complex type such as the ``Movie`` type, which contains both primitive and complex types as properties, MVC's model binding will still handle it nicely. It uses reflection and recursion to traverse the properties of complex types looking for matches. Model binding looks for the pattern `parameter_name.property_name` to perform binding. If it doesn't find matching objects, it will attempt to bind using  just the property name. For complex types such as ``Collection`` types model binding looks for matches to `parameter_name[index]` or just `[index]`. Model binding treats  `Dictionary`` types similarly, asking for `parameter_name[key]` or just `[key]`, as long as they keys are simple types. 
+
+In order for binding to happen, members of complex model types must be public, writable properties containing a default public constructor. Public fields are not bound. If properties are arrays or IEnumerable types, the array properties or types in the IEnumerable must be writable. Model binding handles some objects differently than most. ``ICollection<T>`` is supported for read only properties.  
+
+When a parameter is bound, model binding stops looking for parameters with that name and it moves on to bind the next parameter. If binding fails, MVC does not throw an error. The error is recorded in the ``ModelState.Errors`` collection, and model binding attempts to bind to the next parameter or property. You can query for model state errors by checking the ``ModelState.IsValid`` method. 
+
+.. Note:: The ``ModelState.Errors`` collection is a property of the controller's ``ModelStateDictionary`` object, containing name-value pairs. It's rarely necessary to query this collection yourself; if there are items in this collection, then ``ModelState.IsValid`` returns ``false``. 
+
+Additionally, there are some special data types that MVC must consider when performing model binding:
+
+- ``IFormFile``: Uploaded files that are part of the HTTP request.
```

Comment:
Model binding system also supports writable properties of types that inherit from `IEnumerable<IFormFile>` e.g. `IFormFile[] Files { get; set; }` or `List<IFormFile> Files { get; set; }`.